src/kernel/gmp++/gmp++_int_misc.C

```diff
@@ -392,19 +392,25 @@ namespace Givaro {
     }
     Integer::operator int64_t() const
     {
-#if GMP_LIMB_BITS != 64
-        Integer absThis = abs(*this);
-        int64_t r = static_cast<int64_t>(absThis.operator uint32_t());
-        absThis >>= 32;
-        r |= static_cast<int64_t>(absThis.operator uint32_t()) << 32;
-        return (*this < 0)? -r : r;
+#if __GIVARO_SIZEOF_LONG < 8
+        uint64_t u64 = this->operator uint64_t();
+        /* The following lines are adapted from the file mpz/get_si.c of GMP
+         * source code.
+         */
+        auto sgn = this->sign();
+        if (sgn > 0)
+            return u64 & INT64_MAX;
+        else if (sgn < 0)
+            return -1 - (int64_t) ((u64 -1) & INT64_MAX);
+        else
+            return 0;
 #else
         return mpz_get_si ( (mpz_srcptr)&gmp_rep);
 #endif
     }
     Integer::operator uint64_t() const
     {
-#if GMP_LIMB_BITS != 64
+#if __GIVARO_SIZEOF_LONG < 8
         Integer absThis = abs(*this);
```

Comment:
Don't we want to have a similar treatment for this cast to uint64_t ?

Reply:
No if we want to follow the semantic from GMP. mpz_get_ui does not take into account the sign of the input and return the lowest bits. So to have something similar but for uint64_t, we should just return the lowest 64 bits and forget the sign.